ksql-engine/src/main/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNode.java

```diff
@@ -218,12 +219,53 @@ private static void createSinkTopic(
         : Collections.emptyMap();
 
     kafkaTopicClient.createTopic(kafkaTopicName,
-                                 numberOfPartitions,
-                                 numberOfReplications,
-                                 config
+        numberOfPartitions,
+        numberOfReplications,
+        config
     );
   }
 
+  private SourceTopicProperties getSourceTopicProperties(
+      final String kafkaTopicName,
+      final Map<String, Object> sinkProperties,
+      final KafkaTopicClient kafkaTopicClient
+  ) {
+    // Don't request topic properties from Kafka if both are set in WITH clause.
+    if (sinkProperties.containsKey(KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY)
+        && sinkProperties.containsKey(KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY)) {
+      return new SourceTopicProperties(
+          (Integer) sinkProperties.get(KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY),
+          (Short) sinkProperties.get(KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY)
+      );
+    }
+    final SourceTopicProperties sourceTopicPropertiesFromBroker = getSourceTopicPropertiesFromKafka(
+        kafkaTopicName,
+        kafkaTopicClient);
+
+    final int partitions = (Integer) outputProperties.getOrDefault(
+        KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY,
+        sourceTopicPropertiesFromBroker.getPartitions());
+    final short replicas = (Short) outputProperties.getOrDefault(
+        KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY,
+        sourceTopicPropertiesFromBroker.getReplicas());
+
+    return new SourceTopicProperties(partitions, replicas);
+  }
+
+  private static SourceTopicProperties getSourceTopicPropertiesFromKafka(
+      final String kafkaTopicName,
+      final KafkaTopicClient kafkaTopicClient
+  ) {
+    final Map<String, TopicDescription> topicDescriptionMap = kafkaTopicClient
+        .describeTopics(Collections.singletonList(kafkaTopicName));
+    final TopicDescription topicDescription = topicDescriptionMap.get(kafkaTopicName);
+    Objects.requireNonNull(
+        topicDescription, "Could not fetch source topic description: " + kafkaTopicName);
+    final int partitionCount = topicDescription.partitions().size();
+    final short replicas = (short) topicDescription.partitions().get(0).replicas().size();
+    return new SourceTopicProperties(partitionCount, replicas);
```

Comment:
Just not just return `TopicDescription`  here and delete `SourceTopicProperties`?

Reply:
You're still returning `SourceTopicProperties` from the method.